transport-msg.tex

```diff
@@ -153,40 +153,58 @@ \subsubsection{Transport Revisions and Maximum Message Size}
-in multiple ways:
+by a 16-bit \textbf{device number}. Buses \emph{MAY} discover these device
+numbers in multiple ways:
 \begin{itemize}
   \item \textbf{Message-Based Enumeration}: Using \busref{GET_DEVICES} to query
-        which IDs exist (optional).
+        which numbers exist (optional).
   \item \textbf{Platform Data}: A device tree, ACPI tables, or hypervisor calls
-        might inform the bus of available device IDs and their properties.
+        might inform the bus of available device numbers and their properties.
 \end{itemize}
 
-Once a bus confirms that a given device ID is valid (regardless of the method),
-it typically issues \msgref{GET_DEVICE_INFO} to retrieve the device’s Virtio ID
+Once a bus confirms that a given device number is valid (regardless of the method),
+it typically issues \msgref{GET_DEVICE_INFO} to retrieve the device’s Device ID
 and vendor ID. The bus can then register the device with the host OS to initiate
 the usual Virtio driver binding process.
 
 \subsubsection{Configuration Generation Count}
 \label{sec:Virtio Transport Options / Virtio Over Messages / Basic Concepts / ConfigGen}
 
 Each device maintains a \textbf{Configuration Generation Count} to prevent
-inconsistent updates. This count is incremented whenever the device backend
-changes its own configuration internally. The driver learns the current count
-via \msgref{GET_CONFIG} and \emph{MUST} include it in \msgref{SET_CONFIG}
-requests.
+inconsistent updates. This count is incremented at least once by the device for
+every driver visible change it makes to its own copy of the configuration data.
+A configuration change is made visible to the driver via \msgref{EVENT_CONFIG}
+or the response to \msgref{GET_CONFIG} which also both provide the device's
+current configuration count. The device may change any amount of data for one
+generation count increment. If the change cannot fit in one \msgref{EVENT_CONFIG}
+message, it \emph{SHOULD} be signaled to the driver via a \msgref{EVENT_CONFIG0}
+message. The device \emph{MUST NOT} provide the same generation count in
+multiple \msgref{EVENT_CONFIG} messages that contain non-zero length config
+data. The driver includes its view of the current generation count in
+\msgref{SET_CONFIG} requests. The configuration generation count does not
+necessarily start at 0 and \emph{SHOULD NOT} reset to 0 on device reset.
 
 \begin{itemize}
   \item If the driver’s generation count does not match the device’s current
-        count, the \msgref{SET_CONFIG} request \emph{MUST} be rejected (typically
-        via \msgref{ERROR}).
-  \item The driver then \emph{SHOULD} re-read the configuration (e.g., by
-        \msgref{GET_CONFIG}), retrieve the updated count, and attempt the update
-        again if necessary.
+        count, the \msgref{SET_CONFIG} request \emph{MUST} be rejected by the
+        device.
+  \item The driver will likely have received \msgref{EVENT_CONFIG} messages
```

Comment:
Receiving an event in between a request and its response is something that will only occur in some cases and even though it might be hard for the transport to handle it.
It is a possibility but I would remove the "likely" here and do some rephrasing like:
The driver could receive a msg even config and if not it should re-read the configuration count via get_config.

Reply:
Agreed that the race is unlikely.  I think you understand my description of events.  I will try and reword.

What I was trying to describe:
If the device is following all the recommended SHOULDs (not all in the spec yet) then the EVENT_CONFIG _will_ arrive before the response and the driver can just use the new values and figure out if it still want to write.  

The above in NOT the new words for the spec.  I will work on that.